File: src/native/schemas/desktopApp.ts
import { z } from "zod";
import { desktopAppIdSchema, toolIdSchema } from "./ids";

export { desktopAppIdSchema };

export const desktopAppStatusSchema = z.enum([
  "installed",
  "updateAvailable",
  "notInstalled",
  "unsupported",
  "unknown",
]);

export const desktopAppConfigurationRelationshipSchema = z.enum([
  "sharedConfiguration",
  "separateConfiguration",
  "standaloneApplication",
]);

export const desktopAppInstallerHandoffSchema = z.enum([
  "directOfficialPackage",
  "officialDownloadPage",
  "unsupported",
]);

export const desktopAppUninstallHandoffSchema = z.enum([
  "revealApplication",
  "systemSettings",
  "unsupported",
]);

const desktopVersionSchema = z
  .string()
  .min(1)
  .max(128)
  // eslint-disable-next-line no-control-regex -- control characters are the validation target
  .refine((value) => !/[\u0000-\u001f<>&]/.test(value));

export const desktopAppSchema = z
  .object({
    id: desktopAppIdSchema,
    name: z.string().min(1).max(64),
    status: desktopAppStatusSchema,
    version: desktopVersionSchema.nullable(),
    latestVersion: desktopVersionSchema.nullable().default(null),
    relatedTool: toolIdSchema.nullable(),
    configurationRelationship: desktopAppConfigurationRelationshipSchema,
    canLaunch: z.boolean(),
    environment: z.enum(["macos", "windows", "linux", "unknown"]),
    installerHandoff: desktopAppInstallerHandoffSchema,
    uninstallHandoff: desktopAppUninstallHandoffSchema,
    updatesManagedByVendor: z.boolean(),
    canRollback: z.boolean(),
    canManageMcp: z.boolean(),
  })
  .strict()
  .superRefine((app, context) => {
    const installed =
      app.status === "installed" || app.status === "updateAvailable";
    if (app.canLaunch && !installed) {
      context.addIssue({
        code: "custom",
        message: "only an installed desktop app can be launchable",
      });
    }
    if (
      app.canManageMcp &&
      (app.id !== "claude-desktop" ||
        !installed ||
        (app.environment !== "macos" && app.environment !== "windows"))
    ) {
      context.addIssue({
        code: "custom",
        path: ["canManageMcp"],
        message:
          "only installed Claude Desktop on macOS or Windows can manage MCP",
      });
    }
    const standalone =
      app.configurationRelationship === "standaloneApplication";
    if (standalone !== (app.relatedTool === null)) {
      context.addIssue({
        code: "custom",
        path: ["relatedTool"],
        message: "standalone apps must not claim a related CLI tool",
      });
    }
  });

export const desktopAppListSchema = z
  .array(desktopAppSchema)
  .max(desktopAppIdSchema.options.length)
  .refine(
    (apps) => new Set(apps.map((app) => app.id)).size === apps.length,
    "desktop app ids must be unique",
  );

export const desktopAppLaunchOutcomeSchema = z.literal("launched");
export const desktopAppOfficialDownloadOutcomeSchema = z
  .object({ handoff: desktopAppInstallerHandoffSchema })
  .strict();
export const desktopAppUninstallOutcomeSchema = z.literal("opened");

export type DesktopAppId = z.infer<typeof desktopAppIdSchema>;
export type DesktopAppStatus = z.infer<typeof desktopAppStatusSchema>;
export type DesktopAppConfigurationRelationship = z.infer<
  typeof desktopAppConfigurationRelationshipSchema
>;
export type DesktopAppInstallerHandoff = z.infer<
  typeof desktopAppInstallerHandoffSchema
>;
export type DesktopAppUninstallHandoff = z.infer<
  typeof desktopAppUninstallHandoffSchema
>;
export type DesktopApp = z.infer<typeof desktopAppSchema>;
export type DesktopAppLaunchOutcome = z.infer<
  typeof desktopAppLaunchOutcomeSchema
>;
export type DesktopAppOfficialDownloadOutcome = z.infer<
  typeof desktopAppOfficialDownloadOutcomeSchema
>;
export type DesktopAppUninstallOutcome = z.infer<
  typeof desktopAppUninstallOutcomeSchema
>;

File: src-tauri/src/domain/desktop_app.rs
use serde::{Deserialize, Serialize};

use super::ToolId;

/// Desktop applications are a separate product surface from CLI tools.
/// These identifiers are stable on the native wire and must not be folded into `ToolId`.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum DesktopAppId {
    #[serde(rename = "codex-app")]
    CodexApp,
    #[serde(rename = "claude-desktop")]
    ClaudeDesktop,
    #[serde(rename = "cursor")]
    Cursor,
    #[serde(rename = "zcode")]
    ZCode,
    #[serde(rename = "cherry-studio")]
    CherryStudio,
}

impl DesktopAppId {
    pub const ALL: [Self; 5] = [
        Self::CodexApp,
        Self::ClaudeDesktop,
        Self::Cursor,
        Self::ZCode,
        Self::CherryStudio,
    ];

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::CodexApp => "codex-app",
            Self::ClaudeDesktop => "claude-desktop",
            Self::Cursor => "cursor",
            Self::ZCode => "zcode",
            Self::CherryStudio => "cherry-studio",
        }
    }

    pub fn from_str_id(value: &str) -> Option<Self> {
        Self::ALL.into_iter().find(|id| id.as_str() == value)
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppStatus {
    Installed,
    UpdateAvailable,
    NotInstalled,
    Unsupported,
    Unknown,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppConfigurationRelationship {
    SharedConfiguration,
    SeparateConfiguration,
    StandaloneApplication,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppLaunchOutcome {
    Launched,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppInstallerHandoff {
    DirectOfficialPackage,
    OfficialDownloadPage,
    Unsupported,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppUninstallHandoff {
    RevealApplication,
    SystemSettings,
    Unsupported,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DesktopAppOfficialDownloadOutcome {
    pub handoff: DesktopAppInstallerHandoff,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum DesktopAppUninstallOutcome {
    Opened,
}

/// Safe renderer projection. Bundle paths, package-family names and launch commands
/// remain entirely inside the native platform boundary.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DesktopApp {
    pub id: DesktopAppId,
    pub name: String,
    pub status: DesktopAppStatus,
    pub version: Option<String>,
    pub latest_version: Option<String>,
    pub related_tool: Option<ToolId>,
    pub configuration_relationship: DesktopAppConfigurationRelationship,
    pub can_launch: bool,
    /// Whether this installed desktop configuration surface can safely manage
    /// MCP connections. This is separate from CLI `ToolCapabilities`.
    pub can_manage_mcp: bool,
    pub environment: String,
    pub installer_handoff: DesktopAppInstallerHandoff,
    pub uninstall_handoff: DesktopAppUninstallHandoff,
    pub updates_managed_by_vendor: bool,
    pub can_rollback: bool,
}

#[cfg(test)]
mod tests {
    use super::{
        DesktopApp, DesktopAppConfigurationRelationship, DesktopAppId, DesktopAppInstallerHandoff,
        DesktopAppLaunchOutcome, DesktopAppOfficialDownloadOutcome, DesktopAppStatus,
        DesktopAppUninstallHandoff, DesktopAppUninstallOutcome,
    };
    use crate::domain::ToolId;

    #[test]
    fn desktop_app_ids_are_stable_and_separate_from_cli_ids() {
        for (id, expected) in [
            (DesktopAppId::CodexApp, "codex-app"),
            (DesktopAppId::ClaudeDesktop, "claude-desktop"),
            (DesktopAppId::Cursor, "cursor"),
            (DesktopAppId::ZCode, "zcode"),
            (DesktopAppId::CherryStudio, "cherry-studio"),
        ] {
            assert_eq!(id.as_str(), expected);
            assert_eq!(DesktopAppId::from_str_id(expected), Some(id));
            assert_eq!(
                serde_json::to_string(&id).expect("serialize"),
                format!("\"{expected}\"")
            );
            assert!(ToolId::from_str_id(expected).is_none());
        }
        assert_eq!(DesktopAppId::from_str_id("codex"), None);
    }

    #[test]
    fn desktop_app_projection_round_trips_without_native_paths() {
        let app = DesktopApp {
            id: DesktopAppId::CodexApp,
            name: "ChatGPT / Codex".to_string(),
            status: DesktopAppStatus::Installed,
            version: Some("26.527.60818".to_string()),
            latest_version: Some("26.527.60818".to_string()),
            related_tool: Some(ToolId::Codex),
            configuration_relationship: DesktopAppConfigurationRelationship::SharedConfiguration,
            can_launch: true,
            can_manage_mcp: false,
            environment: "macos".to_string(),
            installer_handoff: DesktopAppInstallerHandoff::DirectOfficialPackage,
            uninstall_handoff: DesktopAppUninstallHandoff::RevealApplication,
            updates_managed_by_vendor: true,
            can_rollback: false,
        };
        let json = serde_json::to_string(&app).expect("serialize app");
        assert_eq!(
            json,
            r#"{"id":"codex-app","name":"ChatGPT / Codex","status":"installed","version":"26.527.60818","latestVersion":"26.527.60818","relatedTool":"codex","configurationRelationship":"sharedConfiguration","canLaunch":true,"canManageMcp":false,"environment":"macos","installerHandoff":"directOfficialPackage","uninstallHandoff":"revealApplication","updatesManagedByVendor":true,"canRollback":false}"#
        );
        assert!(!json.contains("Applications"));
        assert_eq!(
            serde_json::from_str::<DesktopApp>(&json).expect("parse"),
            app
        );
        assert_eq!(
            serde_json::to_string(&DesktopAppLaunchOutcome::Launched).expect("serialize outcome"),
            r#""launched""#
        );
        assert_eq!(
            serde_json::to_string(&DesktopAppOfficialDownloadOutcome {
                handoff: DesktopAppInstallerHandoff::DirectOfficialPackage,
            })
            .expect("serialize official download outcome"),
            r#"{"handoff":"directOfficialPackage"}"#
        );
        assert_eq!(
            serde_json::to_string(&DesktopAppUninstallOutcome::Opened)
                .expect("serialize uninstall outcome"),
            r#""opened""#
        );
    }

    #[test]
    fn every_status_and_relationship_has_a_stable_camel_case_wire_value() {
        for (value, expected) in [
            (DesktopAppStatus::Installed, "installed"),
            (DesktopAppStatus::UpdateAvailable, "updateAvailable"),
            (DesktopAppStatus::NotInstalled, "notInstalled"),
            (DesktopAppStatus::Unsupported, "unsupported"),
            (DesktopAppStatus::Unknown, "unknown"),
        ] {
            assert_eq!(
                serde_json::to_string(&value).expect("serialize"),
                format!("\"{expected}\"")
            );
        }
        assert_eq!(
            serde_json::to_string(&DesktopAppConfigurationRelationship::SeparateConfiguration)
                .expect("serialize"),
            r#""separateConfiguration""#
        );
        assert_eq!(
            serde_json::to_string(&DesktopAppConfigurationRelationship::StandaloneApplication)
                .expect("serialize"),
            r#""standaloneApplication""#
        );
    }
}
